enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,281 @@
+reviewers: 
+  - "@MarSik"
+  - "@yanirq"
+  - "@ffromani"
+  - "@jmencak"
+  - "@mrunalp"
+approvers: 
+  - "@mrunalp"
+api-approvers:
+  - "@mrunalp"
+creation-date: 2023-05-03
+last-updated: 2023-05-08
+tracking-link: 
+  - https://issues.redhat.com/browse/CNF-7603
+---
+
+# mixed-cpu-node-plugin
+
+
+## Summary
+
+Resources management (particularly CPUs) in Kubernetes/OpenShift is limited and not flexible enough to cover all of
+our customer use cases.
+This enhancement introduces a new node-plugin that operates at runtime-level
+for extending CPU resources management on top of Kubernetes and OpenShift platforms.
+With the existing CPU management design, a container can either request exclusive CPUs or shared CPUs,
+while with this new node plugin, it would be possible for container workload to request for both.
+
+## Motivation
+
+There is a growing interest from telco users in Kubernetes/OpenShift as a cloud management platform.
+This proposal is motivated by the desire of optimizing the platform's capabilities for telco users,
+such as cost reduction, power efficiency, higher pod density and sustainability.
+
+The telco users have new use cases and requirements that especially related to resource management of the platform. 
+In order to address some of those needs,
+Kubelet CPU manager was created to allow users requesting for exclusive, isolated cpus.
+
+But there are additional use cases that require more granular control of CPU management.
+This enhancement aims to solve the following use case:
+When the CPU manager allocates an exclusive set of CPUs for a given workloads, the workload shall still have the option 
+to specify and access a set of shared CPUs, whether those shared CPUs are shared among containers in the 
+same pod or containers in different ones.
+
+By giving the users the flexibility to use exclusive CPUs only when it is an absolute necessity, we help them utilize
+their existing compute resources in an optimal manner.
+
+### User Stories
+
+* As a developer who's building DPDK application,
+I want to run my application in a container,
+giving full exclusive access to cpu cores for the busy-loops polling threads,
+but having the light-weight tasks like configuration processing and log printing not taking full cores,
+in order to improve the application density and make more efficient use of the CPUs.
+For that, my application container needs access to both shared and exclusive CPUs.
+
+  
+### Goals 
+* Increase pod density, power efficiency, cost reduction and sustainability by optimizing CPUs resource management. 
+* Provide a special-purpose at a runtime-level solution for containers in a pod to have the ability
+to request and subsequently be allocated both exclusive and shared CPUs.
+* Enable the telco-specific use case with a self-contained change,
+    minimizing the changes to the platform
+
+### Non-Goals
+* Introducing a generic mechanism in the platform that does involve Kubelet and pod spec changes.
```

Comment:
what's the compelling argument for not upstreaming this?

time required?

no one else would use it/be interested in it?

Reply:
I mentioned that in the Alternatives section,
but the gist is that upstreaming this proposal means we need to have a generic solution,
that could fit for broader set of use cases and requirements
and not only the scope-specific requirement that we're trying to address.

In addition, there is an ongoing effort to support greater resource management flexibility upstream (a KEP which is still in discussion) but:
1. We are not sure how another KEP blends alongside this ongoing effort.
2. We can't wait for the existing KEP due to time constraints.  

I mentioned the KEP in the Alternative section as well.

Reply:
Is anyone from the team at least involved in the discussions of that KEP?

If the KEP is approved, how would this design evolve to use it or at least co-exist safely?

Reply:
> Is anyone from the team at least involved in the discussions of that KEP?

Yes. @ffromani and @swatisehgal are.

> If the KEP is approved, how would this design evolve to use it or at least co-exist safely?

The KEP is not going to break any of the existing behavior of CPU manager and this feature meant to work along side CPUManager, hence it won't be affected, by the proposed changes.